### PR TITLE
Refactored Launcher

### LINKED ISSUE #3795: Improve user experience in Launcher
The labels of the python kernels in the launcher are hard to read, especially when you have a large number of kernels. I had to hover over them to read the full name. 

Below is an example of running a Jupyter lab launcher using the Amazon Deep Learning AMI. It would be great to allow a list view and filtering. 

screen shot 2018-02-02 at 10 39 30

### PR DESCRIPTION
**Pull request by zuoyuanh**
Fixes #3795 

In this PR, the new design of Launcher is implemented. This new design is aimed to make kernels with similar names more distinguishable and allow users who have multiple kernels installed to find the kernel they want easier.

**Major Changes**
  - The options for opening a console and creating new notebook for the same kernel are grouped into one single card, with two options
  - Implemented the search feature that allow users to search for a specific kernel or utility.
  - Implemented the usage count for each kernel. Kernels that user used the most will show up in a  new section called 'Most Used', and kernels in that section are sorted by the frequency of usage. 

**Screenshots**

screen shot 2019-02-06 at 5 34 19 pm

### CODE REVIEW

**Comment by ellisonbg** [comment]
Looks like some tests are not passing. @zuoyuanh could you have a look at those

**Comment by ellisonbg** [comment]
Starting review...

* Maybe let's try to use 12px for the name of the kernel.
* When the name of the kernel is really long, it gets hidden under the hover buttons. Not sure of how to handle that. Maybe use "..."
* When I make the JupyterLab window side, the divider borders and cards don't span the whole launcher even though there is room for another column of cards. Looks like `jp-Launcher-content` needs a width of 100% rather than 85%.

screen shot 2019-02-26 at 9 36 02 pm

* I am not able to scroll to the bottom to show the lower part of the last row of cards. I am guessing this has to do with the layout of the toolbar with the search box:

screen shot 2019-02-26 at 9 36 51 pm

* The old launcher had the current path of the file browser listed at the top. I see the DOM node is there, but it isn't visible. Is it underneath the search header. Maybe put the path in the header with the search bar?
* The section header "Environments" should be "Kernels" as environments are a slightly different (but related) abstraction).
* Not sure of the best phrasing for the "Most used" section header. Maybe "Recently Used"? I tested it though and it works nicely ;-) Other ideas? Both words should probably be capitalized.

**Comment by zuoyuanh** [comment]
@ellisonbg Hello Brian, I think I have most of the proposed issues fixed. So for the name of the 'Most used' section, during our design meeting we decided to count how many times each kernel has been used and show the kernel that is used most frequently in this section, and this is why we are using 'Most used' instead of 'Recently used' as the name of this section. Do you think it would be better if we change it to show the most recent used kernels instead of counting the times they are used?

**Comment by ellisonbg** [comment]
@zuoyuanh I will have a look at your updates.

you bring up a really good question about "recent" versus "most used". I see how "most used" is an accurate description of the current algorithm being used.

I am more thinking about the user experience, and maybe we want to use a different usage tracking metric than most used. The problem with most used is that over time, it will become essentially impossible for a users current behavior to change that list. If I open a python kernel 1000 times over 6 months, I am always going to see python in that list, even if I spend a week using only R. I have a feeling this is why many UI's express this as "recently watched/used/opened".

Based on this, maybe calling it "Most Recent" and changing the tracking metric (based on the most recently used item in time) is the best way of doing it? Thoughts?

Cheers,

Brian

**Comment by ellisonbg** [comment]
In local testing, the issues mentioned before seem to be addressed. A few other things:

* I don't think we need to indicate the path with "Current Path:". Something that looks like this:

🏡/Documents/Notebooks

(with a more appropriate "home" icon would probably be clear enough

* I would love to see the path be moved up into the header where the search is, to get the cards higher in the panel. Maybe put the search to the R side of the header, and the path to the L?

**Comment by jasongrout** [comment]
> Based on this, maybe calling it "Most Recent" and changing the tracking metric (based on the most recently used item in time) is the best way of doing it? Thoughts?

Perhaps what is more helpful is some sort of moving average, i.e., one use won't change things, but many uses over a short period of time would?

**Comment by zuoyuanh** [comment]
@ellisonbg Yes I saw that it is not reasonable to compare the usage over all time, and personally I think Jason's idea is way better than the current algorithm

**Comment by ellisonbg** [comment]
I like the time windowing idea as it is a balance between "most recent" and "most used" that probably better represents the users idea of "what was I just working on". So +1 on that, but what to call it?

**Comment by zuoyuanh** [comment]
I feel like probably we can call it 'Frequently Used Recently' or just 'Recently'?

**Comment by jasongrout** [comment]
"Recent" sounds good. It probably makes sense to make *sure* that the last used one is in the list somewhere too, even if it was only used once.

**Comment by weihwang** [comment]
Minor details:

1. Alignment of the cards and the start of the search bar could be flush.
2. If the intention is to not wrap cards to a fixed width, and span the width of the browser, the divider lines should extend as well. Maybe this is already addressed by the 100% width style mentioned earlier by @ellisonbg 
3. General question on language so that the naming is consistent. For "Recent, Environment, Other" Is there something that can be more descriptive for "Other"?

**Comment by ellisonbg** [comment]
@zuoyuanh can you rebase on master?

**Comment by jaipreet-s** [comment]
A minor UX nitpick around ordering of `Console` and `Notebook`: Based on past analysis, I've found that users tend to click the first available option more. In this case, more users would want use the `Notebook` option so it may make sense to be to order it first.

@weihwang: WDYT?

File: packages/launcher/src/index.tsx [hunk 1]
```diff
@@ -7,8 +7,10 @@ import {
   VDomRenderer
 } from '@jupyterlab/apputils';
 
+import { ISettingRegistry } from '@jupyterlab/coreutils';
+
 import {
-  ArrayExt,
+  // ArrayExt,
```
Comment (jaipreet-s) [inline packages/launcher/src/index.tsx]: nitpick: Please remove commented out code before pushing :-)

File: packages/launcher/src/index.tsx [hunk 2]
```diff
@@ -187,25 +381,63 @@ export class Launcher extends VDomRenderer<LauncherModel> {
...
+    let tableCategories = Object.create(null);
+    for (let cat in categories) {
+      tableCategories[cat] = [...categories[cat]];
+    }
+
+    let orderedItems = this.model.sortedItemsByUsage();
+    let i = 0;
+    let cnt = 0;
+    let topUsed: ILauncher.IGroupedItemOptions[] = [];
+    while (cnt < 4 && i < orderedItems.length) {
```
Comment (jaipreet-s) [inline packages/launcher/src/index.tsx]: I didn't get the usage of the hard coded `4` here. Consider adding a comment to document.

File: packages/launcher/src/index.tsx [hunk 3]
```diff
@@ -322,6 +573,75 @@ export namespace ILauncher {
...
+    /**
+     * For items that have a kernel associated with them, the URL of the kernel
+     * icon.
+     *
+     * This is not a CSS class, but the URL that points to the icon in the kernel
+     * spec.
+     */
+    kernelIconUrl?: string;
+
+    mostRecentUsage?: string;
```
Comment (jaipreet-s) [inline packages/launcher/src/index.tsx]: nit: Missing documentation on this field compared to other fields

File: packages/launcher/src/index.tsx [hunk 4]
```diff
@@ -339,50 +659,88 @@ export namespace ILauncher {
...
   kernel: boolean,
-  item: ILauncher.IItemOptions,
+  item: ILauncher.IGroupedItemOptions,
   launcher: Launcher,
   commands: CommandRegistry,
   launcherCallback: (widget: Widget) => void
 ): React.ReactElement<any> {
   // Get some properties of the command
-  const command = item.command;
+  const command = item.commands[item.options[0]];
```
Comment (jaipreet-s) [inline packages/launcher/src/index.tsx]: Why are we only concerned with the 0th option?

File: packages/launcher/src/index.tsx [hunk 5]
```diff
@@ -427,32 +786,86 @@ namespace Private {
...
-      ...options,
-      category: options.category || '',
-      rank: options.rank !== undefined ? options.rank : Infinity
-    };
+  export function getDisplayName(
+    item: ILauncher.IGroupedItemOptions,
+    commands: CommandRegistry,
+    launcher: Launcher
+  ): string {
+    const command = item.commands[item.options[0]];
```
Comment (jaipreet-s) [inline packages/launcher/src/index.tsx]: Same here

File: packages/launcher/src/index.tsx [hunk 6]
```diff
@@ -70,6 +77,21 @@ export interface ILauncher {
...
+ */
+interface IUsageData {
+  /**
+   * Count the number that a certain card is used.
+   */
+  usageCount: number;
+  /**
+   * The most recent timestamp a certain card is used.
+   */
+  mostRecentUsage: string;
```
Comment (jaipreet-s) [inline packages/launcher/src/index.tsx]: any reason to use `Date` vs `string` for the timestamp? (JSON serialization?)

**Comment by tgeorgeux** [comment]
I agree with @jaipreet-s . Putting notebook first makes a lot of sense. The builds are failing for dependency reasons right now, I noticed we had an extra 180k lines of codes from packages that were living inside the origin fork. That's cleaned up but we'll need some time to clean up all the dependencies. 

@zuoyuanh do you mind switching the order of notebook and console if you haven't already?

**Comment by jasongrout** [comment]
We are planning to make a 1.0 release candidate release Saturday morning. What is the status of this PR? Can someone here make a call about how close this is to going in, or what work needs to be done to minimally get it better?

**Comment by ellisonbg** [comment]
@zuoyuanh I will finish this up over the weekend.

**Comment by ellisonbg** [comment]
@zuoyuanh this is fantastic work though and is a huge improvement to the UX.

In doing detailed code review, I think this needs to wait until after 1.0. For 1.0, I will submit a separate PR that does a minor fix of the launcher to better handle long kernel names. I will leave inline comments, but here is an overview:

* This is *almost* API compatible with master, but the `LauncherModel` class has three additional methods that are essentially public by usage, even though they are not in the `ILauncher` interface (`fromJSON`, `useCard`, `setSettingRegistry`).
* The way that individual launcher items are grouped is awkward and we are going to have to break it again in subsequent work on the launcher. The launcher APIs and UX has a clear notion of *category*, but that isn't respected, or at least not in a clear way. Kernels and Consoles are registered with categories that match their name, but then end up appearing in a special "Kernel" category. Where this is going to break down is when we later introduce a tabular view that supports sorting/grouping. In the future API/model, the notebook/console should be registered with a category of "Kernel" and the renderer for that that category should know how to do the grouping of console/kernel within that.
* The search box and cwd UX should be in the toolbar of the activity, and that won't be a purely visual change.
* There are a number of minor UX changes to make, and we could release 1.0 and make those later. I am mostly looking at the bar for 1.0 APIs.

Here is some pseudocode on how I think we should organize the logic for the grouping:

```python
groups = items.groupby('category');
for group in groups:
    if group.name == 'Kernel':
        <KernelCategory group={group}>
    else:
        <GenericCategory group={group}>
```

Where group would be a list of the items in that group. They key is that the grouping logic should be separate from the kernel specific logic, which should be contained in a kernel specific component.

The reason that this will be important is that in a tabular view, we will need to enable the user to group/sort by different columns on the fly.

Where a breaking change will come from in the future is that the notebook and console will need to be added under a "Kernel" category to keep the data model clean.

File: packages/launcher/src/index.tsx [hunk 7]
```diff
@@ -322,6 +572,80 @@ export namespace ILauncher {
      */
     kernelIconUrl?: string;
   }
+
+  export interface IGroupedItemOptions {
```
Comment (ellisonbg) [inline packages/launcher/src/index.tsx]: This interface ends up mixing two ideas: 1) grouping and 2) kernel specific logic. Those two concepts need to be separated, so there is a clear outer loop over groups, and then kernel specific logic for one of the sections (probably a separate `<KernelSection>` component that knows how to render the notebook/console cards).

Comment (ellisonbg) [reply packages/launcher/src/index.tsx]: Also, within the `<KernelSection>` component there should be another groupby on the kernel name that will handle the notebook/console grouping.

File: packages/launcher/src/index.tsx [hunk 8]
```diff
@@ -218,27 +449,46 @@ export class Launcher extends VDomRenderer<LauncherModel> {
...
     return (
-      <div className="jp-Launcher-body">
-        <div className="jp-Launcher-content">
-          <div className="jp-Launcher-cwd">
-            <h3>{this.cwd}</h3>
+      <React.Fragment>
+        <div className="jp-Launcher-body">
+          <div className="jp-Launcher-content">
+            <div className="jp-Launcher-toolbar">
+              <div className="jp-Launcher-search-div">
```
Comment (ellisonbg) [inline packages/launcher/src/index.tsx]: The search and cwd UX should be moved to the actual activity toolbar - here we end up with a double horizontal line above and below this UX.

**Comment by jasongrout** [comment]
Pinging those involved on this PR: what is the current status of this? It looks like great work, but also like there are lots of review comments.

**Comment by ellisonbg** [comment]
@zuoyuanh I am going to close this PR. I think this work still has promise, but I believe we need to look a bit more at how environments are exposed in the launcher.

**Pull request closed without merging**